### PR TITLE
Promote pod readiness gates

### LINKED ISSUE #79115: Promote Pod Test to Conformance
Using the list of tested stable endpoints referenced in https://github.com/kubernetes/kubernetes/issues/78748, we can pick out the endpoints which are part of the Pod kind:

[patchCoreV1NamespacedPodStatus](https://apisnoop.cncf.io/?zoomed=operationId-stable-core-patchCoreV1NamespacedPodStatus&showUntested=false&showConformanceTested=false)

[listCoreV1PodForAllNamespaces](https://apisnoop.cncf.io/?zoomed=operationId-stable-core-listCoreV1PodForAllNamespaces&showUntested=false&showConformanceTested=false)

These are stable core endpoints which have existing testing, potentially making for low-hanging fruit to upgrade to Conformance.

/area test
/area conformance
/sig testing

### PR DESCRIPTION
**Pull request by BobyMCbobs**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Promotes the following E2E test to Conformance:  
test/e2e/common/pods.go: "should support pod readiness gates [NodeFeature:PodReadinessGate]"

**Which issue(s) this PR fixes**:
Fixes #79115

**Special notes for your reviewer**:
Part of Umbrella Issue #78748
Endpoint(s):
- [k8s.io] Pods should support pod readiness gates [NodeFeature:PodReadinessGate]

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```
/area conformance
/area test
@kubernetes/sig-apimachinery-pr-reviews
@kubernetes/sig-architecture-pr-reviews
@kubernetes/cncf-conformance-wg

### CODE REVIEW

File: test/conformance/testdata/conformance.txt [hunk 1]
```diff
@@ -120,6 +120,7 @@ test/e2e/common/pods.go: "should be submitted and removed"
 test/e2e/common/pods.go: "should be updated"
 test/e2e/common/pods.go: "should allow activeDeadlineSeconds to be updated"
 test/e2e/common/pods.go: "should contain environment variables for services"
+test/e2e/common/pods.go: "should support pod readiness gates [NodeFeature:PodReadinessGate]"
```
Comment (johnbelamaric) [inline test/conformance/testdata/conformance.txt]: I believe the tooling strips all tags, so the tag should be taken out.

File: test/e2e/apimachinery/chunking.go [hunk 2]
```diff
@@ -75,7 +75,12 @@ var _ = SIGDescribe("Servers with support for API chunking", func() {
 		})
 	})
 
-	ginkgo.It("should return chunks of results for list calls", func() {
+	/*
+		Release : v1.15
```
Comment (johnbelamaric) [inline test/e2e/apimachinery/chunking.go]: Looks like you left this in from another PR?

File: test/e2e/common/pods.go [hunk 3]
```diff
@@ -771,8 +771,13 @@ var _ = framework.KubeDescribe("Pods", func() {
 		}
 	})
 
+	/*
+		Release : v1.15
```
Comment (johnbelamaric) [inline test/e2e/common/pods.go]: should be v1.16

File: test/e2e/common/pods.go [hunk 4]
```diff
@@ -771,8 +771,13 @@ var _ = framework.KubeDescribe("Pods", func() {
...
+		Testname: Ensure readiness gates are respected
+		Description: Create a pod with multiple readiness gates and ensure the pod is unReady until all gates are ready
```
Comment (johnbelamaric) [inline test/e2e/common/pods.go]: Rewrite this in more spec-y language:

Create a pod with multiple readiness gates. The pod MUST become ready only after all gates are ready.

File: test/e2e/common/pods.go [hunk 5]
```diff
@@ -771,8 +771,13 @@ var _ = framework.KubeDescribe("Pods", func() {
...
+	*/
 	// TODO(freehan): label the test to be [NodeConformance] after tests are proven to be stable.
-	ginkgo.It("should support pod readiness gates [NodeFeature:PodReadinessGate]", func() {
+	framework.ConformanceIt("should support pod readiness gates [NodeFeature:PodReadinessGate]", func() {
```
Comment (johnbelamaric) [inline test/e2e/common/pods.go]: I think you should use the agnhost image with the pause command.

File: test/e2e/common/pods.go (hunk 4 above)
Comment (hh) [reply test/e2e/common/pods.go]: RFC2119 style with CAPS or not :smile: ?
```
The pod MUST become ready ONLY after all gates are ready.
```

File: test/e2e/common/pods.go (hunk 5 above)
Comment (hh) [reply test/e2e/common/pods.go]: @bobymcbobs something like this:

```go
Spec: v1.PodSpec{
  TerminationGracePeriodSeconds: &immediate,
    Containers: []v1.Container{
{
  Name:  "agnhost-pause",
  Image: imageutils.GetE2EImage(imageutils.Agnhost),
  Args:  []string{"pause"},
 },
```

File: test/e2e/apimachinery/chunking.go (hunk 2 above)
Comment (devaips) [reply test/e2e/apimachinery/chunking.go]: I've commited the chunking.go file without the changes from the other PR. Let's make sure to rebase before this is merged. cc @BobyMCbobs @hh

**Comment by devaips** [comment]
Let's be sure to rebase before this is merged to remove the rogue commit from the other PR.

File: test/e2e/common/pods.go (hunk 5 above)
Comment (BobyMCbobs) [reply test/e2e/common/pods.go]: @hh and @johnbelamaric, I've swapped out the busybox image for Agnhost.

File: test/e2e/common/pods.go (hunk 4 above)
Comment (BobyMCbobs) [reply test/e2e/common/pods.go]: The description has been updated to be to the point.

File: test/e2e/common/pods.go [hunk 6]
```diff
@@ -771,8 +771,15 @@ var _ = framework.KubeDescribe("Pods", func() {
...
 	})
 
+	/*
+				Release : v1.16
+				Testname: Ensure readiness gates are respected
+				Description: Create a pod with two readiness test conditions. Submit the pod to Kubernetes, checking the readiness MUST return false at this point.
+		        After waiting for a 10 seconds the pod's readiness state MUST return true.
+		        Finally, the state will be checked again, the state of the pod's readiness MUST NOT return false.
+	*/
 	// TODO(freehan): label the test to be [NodeConformance] after tests are proven to be stable.
```
Comment (johnbelamaric) [inline test/e2e/common/pods.go]: go ahead and do this TODO, assuming this is not flaky

File: test/e2e/common/pods.go (hunk 5 above)
Comment (johnbelamaric) [reply test/e2e/common/pods.go]: I can't bear having a "sleep" in a test like this. It just screams "one day I will become flaky". Can we change the sleep to blocking on a watch with a 20 second timeout or something similar?

**Review by johnbelamaric — CHANGES_REQUESTED** [review]
I think we need to get rid of the "sleep" before we can be confident this won't become flaky at some point in the future.

File: test/e2e/common/pods.go (hunk 5 above)
Comment (hh) [reply test/e2e/common/pods.go]: Do we have a recommended way to perform watches within conformance tests?
I've seen several approaches, and might be nice to have a preferred one pointed out.

Do we want to create the re-usable wait function ala:
https://github.com/kubernetes/kubernetes/pull/79667/files#diff-b017c204d2b1afdc6ed2c313afaffcdcR460

Or should we write some small code wrapping a wait.

**Comment by devaips** [comment]
Closing per comments above. @BobyMCbobs and I are working on a rewrite.

/close

**Pull request closed without merging**